packages/clerk-js/src/ui/hooks/useAssurance.ts

```diff
@@ -0,0 +1,145 @@
+import { ClerkRuntimeError, isClerkAPIResponseError } from '@clerk/shared/error';
+import { useClerk } from '@clerk/shared/react';
+import type { Clerk } from '@clerk/types';
+import { useMemo } from 'react';
+
+interface PromiseWithResolvers<T> {
+  promise: Promise<T>;
+  resolve: (value: T | PromiseLike<T>) => void;
+  reject: (reason?: any) => void;
+}
+
+type AssuranceHint = {
+  clerk_error: {
+    type: 'forbidden';
+    reason: 'assurance';
+  };
+};
+
+function generateAssuranceHint<Metadata extends Record<string, string>>(metadata?: Metadata): AssuranceHint;
+function generateAssuranceHint<Metadata extends Record<string, string>>(metadata: Metadata, as: 'response'): Response;
+function generateAssuranceHint<Metadata extends Record<string, string>>(
+  metadata?: Metadata,
+  as?: 'response',
+): AssuranceHint | Response {
+  const obj = {
+    clerk_error: {
+      type: 'forbidden',
+      reason: 'assurance',
+      metadata,
+    },
+  } as AssuranceHint;
```

Comment:
Why not add the `metadata` to the `AssuranceHint`, then you don't need to typecast

Reply:
I may end up changing this in a followup iteration.